File: src/main/java/net/katsstuff/danmakucore/network/ParticlePacket.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.network;

import io.netty.buffer.ByteBuf;
import net.katsstuff.danmakucore.client.particle.GlowTexture;
import net.katsstuff.danmakucore.client.particle.ParticleUtil;
import net.katsstuff.danmakucore.data.Vector3;
import net.minecraft.client.Minecraft;
import net.minecraft.network.PacketBuffer;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;

public class ParticlePacket {

	public static class Message implements IMessage {

		private Vector3 pos;
		private Vector3 motion;
		private float r;
		private float g;
		private float b;
		private float scale;
		private int lifetime;
		private GlowTexture type;

		public Message(Vector3 pos, Vector3 motion, float r, float g, float b, float scale, int lifetime, GlowTexture type) {
			this.pos = pos;
			this.motion = motion;
			this.r = r;
			this.g = g;
			this.b = b;
			this.scale = scale;
			this.lifetime = lifetime;
			this.type = type;
		}

		public Message() {}

		@Override
		public void fromBytes(ByteBuf byteBuf) {
			PacketBuffer buf = new PacketBuffer(byteBuf);
			pos = new Vector3(buf.readDouble(), buf.readDouble(), buf.readDouble());
			motion = new Vector3(buf.readDouble(), buf.readDouble(), buf.readDouble());
			r = buf.readFloat();
			g = buf.readFloat();
			b = buf.readFloat();
			scale = buf.readFloat();
			lifetime = buf.readInt();
			type = buf.readEnumValue(GlowTexture.class);
		}

		@Override
		public void toBytes(ByteBuf byteBuf) {
			PacketBuffer buf = new PacketBuffer(byteBuf);
			buf.writeDouble(pos.x());
			buf.writeDouble(pos.y());
			buf.writeDouble(pos.z());
			buf.writeDouble(motion.x());
			buf.writeDouble(motion.y());
			buf.writeDouble(motion.z());
			buf.writeFloat(r);
			buf.writeFloat(g);
			buf.writeFloat(b);
			buf.writeFloat(scale);
			buf.writeInt(lifetime);
			buf.writeEnumValue(type);
		}
	}

	public static class Handler implements IMessageHandler<Message, IMessage> {

		@Override
		public IMessage onMessage(Message message, MessageContext ctx) {
			Minecraft.getMinecraft().addScheduledTask(
					() -> ParticleUtil.spawnParticleGlow(Minecraft.getMinecraft().world, message.pos, message.motion, message.r, message.g,
							message.b,
							message.scale, message.lifetime, message.type));
			return null;
		}
	}
}

File: src/main/scala/net/katsstuff/danmakucore/network/ParticlePacket.scala
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.network

import net.katsstuff.danmakucore.client.particle.{GlowTexture, ParticleUtil}
import net.katsstuff.danmakucore.data.Vector3
import net.katsstuff.danmakucore.network.scalachannel.{ClientMessageHandler, MessageConverter}
import net.minecraft.client.Minecraft
import net.minecraft.client.network.NetHandlerPlayClient

case class ParticlePacket(
    pos: Vector3,
    motion: Vector3,
    r: Float,
    g: Float,
    b: Float,
    scale: Float,
    lifetime: Int,
    texture: GlowTexture
)
object ParticlePacket {

  implicit val converter: MessageConverter[ParticlePacket] = MessageConverter.mkDeriver[ParticlePacket].apply

  implicit val handler: ClientMessageHandler[ParticlePacket, Unit] = new ClientMessageHandler[ParticlePacket, Unit] {
    override def handle(netHandler: NetHandlerPlayClient, a: ParticlePacket): Option[Unit] = {
      scheduler.addScheduledTask(ParticlePacketRunnable(netHandler, a))
      None
    }
  }
}

case class ParticlePacketRunnable(server: NetHandlerPlayClient, packet: ParticlePacket) extends Runnable {
  override def run(): Unit =
    ParticleUtil.spawnParticleGlow(
      Minecraft.getMinecraft.player.world,
      packet.pos,
      packet.motion,
      packet.r,
      packet.g,
      packet.b,
      packet.scale,
      packet.lifetime,
      packet.texture
    )
}
